client/src/language/codeActionProvider.ts

```diff
@@ -0,0 +1,101 @@
+
+export class BitbakeCodeActionProvider implements vscode.CodeActionProvider {
+  async provideCodeActions (document: vscode.TextDocument, range: vscode.Range | vscode.Selection, context: vscode.CodeActionContext, token: vscode.CancellationToken): Promise<vscode.CodeAction[]> {
+    const diagnostics = context.diagnostics
+    const actions = await Promise.all(
+      diagnostics.map(async (diagnostic) => await buildActionFromDiagnostic(document, diagnostic))
+    ).then((results) => results.flat())
+    return actions
+  }
+}
+
+const buildActionFromDiagnostic = async (document: vscode.TextDocument, diagnostic: vscode.Diagnostic): Promise<vscode.CodeAction[]> => {
+  const originalRange = diagnostic.range
+  const embeddedLanguageType = await requestsManager.getEmbeddedLanguageTypeOnPosition(document.uri.toString(), originalRange.start)
+  if (embeddedLanguageType === undefined || embeddedLanguageType === null) {
+    return []
+  }
+  const embeddedLanguageDocInfos = embeddedLanguageDocsManager.getEmbeddedLanguageDocInfos(document.uri.toString(), embeddedLanguageType)
+  if (embeddedLanguageDocInfos === undefined || embeddedLanguageDocInfos === null) {
+    return []
+  }
+  const embeddedLanguageTextDocument = await vscode.workspace.openTextDocument(embeddedLanguageDocInfos.uri)
+  const embeddedRange = getEmbeddedLanguageDocRange(
+    document,
+    embeddedLanguageTextDocument,
+    embeddedLanguageDocInfos.characterIndexes,
+    originalRange
+  )
+
+  const tempActions = await vscode.commands.executeCommand<vscode.CodeAction[]>(
+    'vscode.executeCodeActionProvider',
+    embeddedLanguageDocInfos.uri,
+    embeddedRange
+  )
+
+  const actions: vscode.CodeAction[] = []
+  tempActions.forEach((action) => {
+    switch (action.command?.command) {
+      case 'python.addImport':
+        handlePythonAddImport(action, document, embeddedLanguageTextDocument, embeddedLanguageDocInfos.characterIndexes)
+        break
+      default:
+        return
+    }
+    actions.push(action)
+  })
+
+  return actions
+}
+
+type PythonAddImportArguments = [
+  path: string,
+  range: Range,
+  targetName: string,
+  moduleName: string | null,
+]
+const handlePythonAddImport = (
+  action: vscode.CodeAction,
+  originalTextDocument: vscode.TextDocument,
+  embeddedLanguageTextDocument: vscode.TextDocument,
+  characterIndexes: number[]
+): void => {
+  if (action.command === undefined) {
+    return
+  }
+  if (action.command?.command !== 'python.addImport') {
+    logger.error(`[handlePythonAddImport] Invalid command ${action.command?.command} (should be 'python.addImport')`)
+    return
+  }
+  const [, range, targetName, moduleName] = action.command?.arguments as PythonAddImportArguments
+  const originalStartPosition = getOriginalDocPosition(
+    originalTextDocument,
+    embeddedLanguageTextDocument,
+    characterIndexes,
+    new vscode.Position(range.start.line, range.start.character)
+  )
+  if (originalStartPosition === undefined) {
```

Comment:
you could consider using exceptions accross this and getOriginalDocPosition to make error handling less lines of code and more readable.

Reply:
you'd like getOriginalDocPosition to throw an exception? I think it would be more error prone, since we could forget to handle the exception. Plus we would have to deal with "unknown".